macaroons/constraints.go

```diff
@@ -87,3 +115,90 @@ func IPLockChecker() (string, checkers.Func) {
 		return nil
 	}
 }
+
+// CustomConstraint returns a function that adds a custom caveat condition to
+// a macaroon.
+func CustomConstraint(name, condition string) func(*macaroon.Macaroon) error {
+	return func(mac *macaroon.Macaroon) error {
+		if name != "" {
+			// The inner (custom) condition is optional.
+			outerCondition := fmt.Sprintf("%s %s", name, condition)
+			if condition == "" {
+				outerCondition = name
+			}
+
+			caveat := checkers.Condition(
+				CondLndCustom, outerCondition,
+			)
+			return mac.AddFirstPartyCaveat([]byte(caveat))
+		}
+
+		return nil
+	}
+}
+
+// CustomChecker returns a Checker function that is used by the macaroon bakery
+// library to check whether a custom caveat is supported by lnd in general or
+// not. Support in this context means: An additional gRPC interceptor was set up
+// that validates the content (=condition) of the custom caveat. If such an
+// interceptor is in place then the acceptor should return a nil error. If no
+// interceptor exists for the custom caveat in the macaroon of a request context
+// then a non-nil error should be returned and the macaroon is rejected as a
+// whole.
+func CustomChecker(acceptor CustomCaveatAcceptor) Checker {
+	// We return the general name of all lnd custom macaroons and a function
+	// that splits the outer condition to extract the name of the custom
+	// condition and the condition itself. In the bakery library that's used
+	// here, a caveat always has the following form:
+	//
+	// <condition-name> <condition-value>
+	//
+	// Because a checker function needs to be bound to the condition name we
+	// have to choose a static name for the first part ("lnd-custom", see
+	// CondLndCustom. Otherwise we'd need to register a new Checker function
+	// for each custom caveat that's registered. To allow for a generic
+	// custom caveat handling, we just add another layer and expand the
+	// initial <condition-value> into
+	//
+	// "<custom-condition-name> <custom-condition-value>"
+	//
+	// The full caveat string entry of a macaroon that uses this generic
+	// mechanism would therefore look like this:
+	//
+	// "lnd-custom <custom-condition-name> <custom-condition-value>"
+	checker := func(_ context.Context, _, outerCondition string) error {
+		outerCondition = strings.TrimSpace(outerCondition)
+		if outerCondition == "" {
+			return fmt.Errorf("expected custom caveat, got empty " +
+				"string")
+		}
+
+		// The condition part of the original caveat is now name and
+		// condition of the custom caveat (we add a layer of conditions
+		// to allow one custom checker to work for all custom lnd
+		// conditions that implement arbitrary business logic).
+		parts := strings.Split(outerCondition, " ")
```

Comment:
No, I think we don't want to restrict what the condition should look like. It might contain some spaces itself, which is fine, as long as we can extract a name from the first part.